jax_rl/MPO.py

```diff
@@ -0,0 +1,379 @@
+            sig_lagrange_params
+        )
+        self.sig_lagrange_optimizer = jax.device_put(sig_lagrange_optimizer)
+
+        self.eps_q = eps_q
+        self.eps_mu = eps_mu
+        self.eps_sig = eps_sig
+        self.temp_steps = temp_steps
+
+        self.max_action = max_action
+        self.discount = discount
+        self.target_freq = target_freq
+
+        self.action_dim = action_dim
+
+        self.total_it = 0
+
+    @property
+    def td_params(self):
+        return self.discount, self.max_action, self.actor_target, self.critic_target
+
+    @property
+    def e_params(self):
+        return (
+            self.actor_target_params,
+            self.critic_target_params,
+            self.max_action,
+            self.action_dim,
+            self.temp_optimizer,
+            self.eps_q,
+            self.temp_steps,
+        )
+
+    @property
+    def m_params(self):
+        return (
+            self.actor_optimizer,
+            self.actor_target_params,
+            self.eps_mu,
+            self.eps_sig,
+            self.mu_lagrange_optimizer,
+            self.sig_lagrange_optimizer,
+        )
+
+    def select_action(self, state):
+        mu, _ = apply_model(actor, self.actor_optimizer.target, state.reshape(1, -1))
+        return mu
+
+    def sample_action(self, rng, state):
+        mu, log_sig = apply_model(
+            actor, self.actor_optimizer.target, state.reshape(1, -1)
+        )
+        sig = jnp.abs(log_sig)
+        return mu + random.normal(rng, mu.shape) * sig
+
+    def train(self, replay_buffer, batch_size, action_sample_size):
+        self.total_it += 1
+
+        buffer_out = replay_buffer.sample(next(self.rng), batch_size)
+
+        target_Q = jax.lax.stop_gradient(
+            get_td_target(next(self.rng), *buffer_out, *self.td_params)
+        )
+
+        state, action, *_ = buffer_out
+
+        self.critic_optimizer = critic_step(
+            self.critic_optimizer, state, action, target_Q
+        )
+
+        self.temp_optimizer, weights, sampled_actions = e_step(
+            next(self.rng), *self.e_params, state, batch_size, action_sample_size
+        )
+
+        weights, sampled_actions = list(
+            map(jax.lax.stop_gradient, [weights, sampled_actions])
+        )
+        sampled_actions = (
+            sampled_actions.reshape((batch_size, action_sample_size, self.action_dim))
+            .transpose((0, 1))
+            .squeeze()
+        )
+
+        rngs = [next(self.rng) for _ in range(3)]
+
+        (
+            self.mu_lagrange_optimizer,
+            self.sig_lagrange_optimizer,
+            self.actor_optimizer,
+        ) = m_step(rngs, *self.m_params, state, weights, sampled_actions)
+
+        if self.total_it % self.target_freq == 0:
+            self.actor_target_params = self.actor_target.replace(
+                params=self.actor_optimizer.target.params
+            )
+            self.critic_target_params = self.critic_target.replace(
```

Comment:
I think this is wrong. You are updating the optimizer instead of taking paramaters from the optimizers. Here `self.critic_target_param` becomes an `Optimizer`